### PR TITLE
Coverage

### PR DESCRIPTION
**Pull request by sullof**
__Test coverage__

`yarn run test-src-web`
runs tests against the source code and generates a coverage. The files are transpiled at runtime and executed it in a Chrome Headless environment.

`yarn run test-dist-node`
runs test against the transpiled code.

--------------------
_Original description (to understand the comments)_

This adds coverage with reports to see what is missed.
It uses Karma and is executing the tests in a Chrome Headless instance. So, in the future, we should be able to perform web tests (React, etc...).
I could handle some transformations. So, I moved statements like
```
import providers from 'lib/providers';
```
to
```
import providers from './lib/providers';
```
Also, I added a 
```
} catch (err) {
```
to where it was 
```
} catch {
```
Finally, I also removed
```
static providers = providers
```
and added at the end
```
TronWeb.providers = providers
```
In a future version we can reuse those experimental statements (I like them).

The tests must mirror the structure of the `src` folder and have the suffix `_test`. For example, to test `src/lib/trx.js` we have to add the file `test/lib/trx_test.js`.

### CODE REVIEW

**Comment by sullof** [comment]
I also changed 
```
feeLimit = 1_000_000_000,
```
to
```
feeLimit = 1000000000,
```
to make it compatible with the `istanbul-instrumenter-loader`.

**Comment by sullof** [comment]
@Kondax @kookiekrak Before merging it, please, verify that it works. Unfortunately here in the coworking I am still waiting for the net admin to allow me to access the not-standard https ports (like 8090, 8091, etc.), so I cannot verify if the example works.

**Comment by Kondax** [comment]
Is there any way to use Babel with karma so we don't lose what you've mentioned above?

**Comment by sullof** [comment]
I tried but it didn't work. I can give it another try.

**Comment by sullof** [comment]
So, I tried to use `babel-loader` first and `istanbul-instrumenter-loader` post. But the coverage is not generated correctly and all the files are empty. I think that `istanbul-instrumenter-loader` uses `babel-loader` under the surface, but uses the version `^7.1.2` instead of `^8.0.2` we uses. I think than soon they will update it and we will be able to reverse the changes.

**Comment by sullof** [comment]
I tried before to compile for node, but I was unable to find a way to have a coverage and I ended up using Karma, which works with a virtual browser.
I can go back to my previous attempt and try to make it working.

**Comment by Kondax** [comment]
If possible we should target node as they're separate browsers (and
separate builds).

We should also test the built files along with the source files - they can
act differently if they're transpiled.

**Comment by sullof** [comment]
I agree.
Testing the transpiled files is trivial and can be easily managed with a ENV variable, even if we cannot test everything there, because the components are not exposed in the dist files.

I am focused on the node coverage. Hope that I can make it soon.

**Comment by Kondax** [comment]
I've had a look and it might be a better idea to use babel-plugin-istanbul
instead of instanbul-instrument-loader.

It's more up to date and supports Babel 7.

**Comment by sullof** [comment]
You were right. `babel-plugin-istanbul` supports all the statements.
Also, I readded the test of the dist-node files. The solution for now is not the best, but it is hard to share the files among the two tests because one is tested in node and the other is tested in the browser. So, env variables, conf files, etc. don't work the same way in the two environments. So, for now, I just duplicated the files. I will think a better way to avoid code repetitions in a future PR.

**Comment by Kondax** [comment]
Hmm, so karma cannot run tests for both NodeJS and Chrome headless, only one? We might want to try and rectify this so it doesn't become a mess in the future.

**Comment by sullof** [comment]
I tried to add node test without success in another branch. In this PR I am testing the dist files, i.e., the original test, just moving it in a subfolder to avoid conflicts.

File: test/dist-node/index-spec.js [hunk 1]
```diff
@@ -239,6 +239,15 @@ describe('TronWeb Instance', () => {
 
             assert.equal(isConnected, false);
         });
+
+        it('should return false if the eventServer is null', async () => {
```
Comment (Kondax) [inline test/dist-node/index-spec.js]: We shouldn't directly modify `tronWeb.eventServer` (and I'd like to disable this through a proxy object). Instead you should even create an instance without an event server specified or call `tronWeb.setEventServer()` which will default to `false`.

File: test/src-web/src/index_test.js [hunk 2]
```diff
@@ -239,6 +239,15 @@ describe('TronWeb Instance', () => {
 
             assert.equal(isConnected, false);
         });
+
+        it('should return false if the eventServer is null', async () => {
```
Comment (Kondax) [inline test/src-web/src/index_test.js]: Same issue as mentioned above

File: test/dist-node/index-spec.js [hunk 3]
```diff
@@ -350,4 +359,16 @@ describe('TronWeb Instance', () => {
             assert.equal(providers.eventServer, EVENT_API);
         });
     });
+
+    describe('#currentProvider()', () => {
+        it('should return the current providers', () => {
```
Comment (Kondax) [inline test/dist-node/index-spec.js]: Do we need to test this? It's an alias of a function so does not actually do anything itself.

File: test/src-web/src/index_test.js [hunk 4]
```diff
@@ -350,4 +359,16 @@ describe('TronWeb Instance', () => {
             assert.equal(providers.eventServer, EVENT_API);
         });
     });
+
+    describe('#currentProvider()', () => {
+        it('should return the current providers', () => {
```
Comment (Kondax) [inline test/src-web/src/index_test.js]: Same as above

**Comment by Kondax** [comment]
What are the conflicts? We should be able to run both tests with the same test files.

File: test/dist-node/index-spec.js (hunk 3 above)
Comment (sullof) [reply test/dist-node/index-spec.js]: I agree it is not really useful, but I add to have a better coverage.

File: test/src-web/src/index_test.js (hunk 2 above)
Comment (sullof) [reply test/src-web/src/index_test.js]: Same response as above, it is not really useful, but decreases the coverage if not.

**Comment by sullof** [comment]
The way you load TronWeb is very different in the two environments. In the browser the entire object has been transpiled previously by Webpack, in node the object is required a runtime. There is no way to put there a condition and decide what to load based on the environment. At least, I haven't found one. I will think about it.

**Comment by Kondax** [comment]
I was just looking for the best approach for doing the above and I've come across https://github.com/karma-runner/grunt-karma. We'd provide a base config file, and then two specific configs which are merged with the base config (one for node, one for browsers).

**Comment by sullof** [comment]
It looks interesting. I was thinking of building a very simple automation tool to compose the test files before running the tests. This is basically the same approach.

**Comment by sullof** [comment]
Testing on the other browsers you mentioned is feasible, since there is a karma launcher for all of them.

**Comment by sullof** [comment]
@Kondax 
I added tests for Firefox and Edge. They work only if you have the browser installed on your machine. In my case, since I have a macbook, I don't have Edge. But the task is completed anyway ignoring the missing browser.
I have also added an initial configuration for grunt-karma.
Tomorrow morning I will continue.

**Comment by Kondax** [comment]
Sounds good. Since this is a public repo (assuming you / Tian have appropriate permission) do we want to benefit from adding Travis or another CI for automatically running tests on PRs?

**Comment by kookiekrak** [comment]
Works for me, no time right now to set it up though

**Comment by sullof** [comment]
`grunt-karma` allows to create multiple Karma configurations, but Karma works only for tests in browsers, so we cannot use it to test the code against Node.

I am simply using karma for tests in browsers to generate a coverage report. Separately, we run the tests against Node, starting from the transpiled code.
For Node tests, we don't need coverage (since we are going to cover the same cases). 

About using Travis, a simpler solution, at least for now, is to force tests before pushing. I added a `pre-push` script in git to force successful tests before pushing a change. It is more common to use a `pre-commit`, but this does not allow you to commit temporary changes if the tests fail, and sometimes that is very useful. Anyway, if there are unwanted commits, we can rebase.

Also, since Mocha uses its own context, it is a best practice to not use arrows in tests, but use `function`, so that the right context can be kept. In our specific case, this is important because the test `#isEventServerConnected()` often fails in some browsers (mostly Firefox) due to a timeout.
So, I set a different timeout only for that case.

Finally, I have left, for now, the Grunt-karma config in case we decide to use it anyway.

I think that this is a good starting point. What do you think?

**Comment by sullof** [comment]
I also added in pre-push a check to block pushing to master. Of course, someone can change these rules before pushing, and this is why it would be better to use Travis CI or something similar, but for now, it is good, at least, to avoid pushing to master by mistake.

**Comment by Kondax** [comment]
Seems like a good compromise. Regarding blocking master commits maybe Tian
could disable this in the repository settings?

For the pre-push and pre-commit checks do they only run on *nix? We need to
support Windows too, we could simply use a node script instead of bash
script to check the tests.

**Comment by sullof** [comment]
I agree on disabling push to master in the settings. @kookiekrak?

About the pre-push, you are right, it can work on Windows 10 under some circumstances, but it won't work on previous versions. I suggest we fix this in a future PR, though, since I can't do it now because I am working on Tronbox.

**Comment by Kondax** [comment]
I'll be able to fix it, I'll do so later.

**Comment by kookiekrak** [comment]
i'll disable, need to get admin access first

**Comment by Kondax** [comment]
Changes made in #20 
Test suite now supports multi-os.

**Pull request merged**